payment/key.go

```diff
@@ -73,7 +84,131 @@ func GenerateSecret() (secret string, nonce string, err error) {
 	if err != nil {
 		return "", "", err
 	}
+	unencryptedSecret = SecretTokenPrefix + unencryptedSecret
+
 	encryptedBytes, nonceBytes, err := cryptography.EncryptMessage(byteEncryptionKey, []byte(unencryptedSecret))
 
 	return fmt.Sprintf("%x", encryptedBytes), fmt.Sprintf("%x", nonceBytes), err
 }
+
+// LookupVerifier returns the merchant key corresponding to the keyID used for verifying requests
+func (service *Service) LookupVerifier(ctx context.Context, keyID string) (context.Context, *httpsignature.Verifier, error) {
+	rootKeyIDStr, caveats, err := cryptography.DecodeKeyID(keyID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rootKeyID, err := uuid.FromString(rootKeyIDStr)
+	if err != nil {
+		return nil, nil, fmt.Errorf("root key id must be a uuid: %v", err)
+	}
+
+	key, err := service.Datastore.GetKey(rootKeyID, false)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	secretKey, err := key.GetSecretKey()
+	if err != nil {
+		return nil, nil, err
+	}
+	if secretKey == nil {
+		return nil, nil, errors.New("missing secret key")
+	}
+
+	secretKeyStr := *secretKey
+
+	if caveats != nil {
+		_, secretKeyStr, err = cryptography.Attenuate(rootKeyID.String(), secretKeyStr, caveats)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		ctx = context.WithValue(ctx, caveatsCtxKey{}, caveats)
+	}
+
+	ctx = context.WithValue(ctx, merchantCtxKey{}, key.Merchant)
+
+	verifier := httpsignature.Verifier(httpsignature.HMACKey(secretKeyStr))
+	return ctx, &verifier, nil
+}
+
+// MerchantSignedMiddleware requires that requests are signed by valid merchant keys
+func (service *Service) MerchantSignedMiddleware() func(http.Handler) http.Handler {
+	merchantVerifier := httpsignature.ParameterizedKeystoreVerifier{
+		SignatureParams: httpsignature.SignatureParams{
+			Algorithm: httpsignature.HS2019,
+			Headers: []string{
+				"(request-target)", "host", "date", "digest",
```

Comment:
I think we should also include `content-length` and `content-type` in the signature just to add another layer of defenses against tampering with non-signed portions of the request.